Style/Frontend/index.md

```diff
@@ -0,0 +1,1451 @@
+
+## Hooks
+
+From the React Native Docs: "**What is a Hook?** A Hook is a special function that lets you “hook into” React features. For example, `useState` is a Hook that lets you add React state to function components."
+
+### [useState](https://reactjs.org/docs/hooks-state.html)
+
+You may be thinking, 'if all of our components are functions, how is information preserved between function calls?'. The answer is state. Redux Toolkit allows us to manage state independently of our components, and lets us interact with the state via a single mutable reference (which is `state`, that thing passed to every reducer and selector), and that seems all fine but what if we need something outside of that system, say to store a unique error message or loading state in only one specific component? It doesn't make sense to add something to redux unless multiple parts of the application depend on that state, so if we want a component to be stateful without Redux Toolkit, we need `useState`.
+
+```jsx
+// The state example from the docs
+const Widget = ({}) => {
+  // useState returns the reference to the state variable and the function to manipulate it safely
+  // You can rename these to whatever, but since we are making a counter we have `count` and `setCount`
+  const [count, setCount] = useState(0);
+  // Use the given function to manipulate the state
+  const increment = () => setCount(count + 1);
+
+	return (
+    // Now every time we press this button, our state changes which triggers a rerender that then shows the new changes
+    <Button onClick={increment}>
+    	I count things
+    </Button>
+  );
+}
+```
+
+### [useCallback](https://reactjs.org/docs/hooks-reference.html#usecallback)
+
+When you define something in a component, it is created on each render. What `useCallback` does is it lets you define something, and it only gets redefined if the dependencies of that thing changes. What this means is it lets you memoize some arbitrary callback, so very useful if you have a subcomponent defined in the main component or if you have a very heavy function that shouldn't be ran on each render.
+
+```jsx
+const foo = (data) => myExpensiveOperation(data); // gets created every render
+
+// callbackized version
+const foo = useCallback(
+  () => myExpensiveOperation(data),
+  [data], // dependency array, means this only runs again if data changes
+);
+
+// note, some things are NOT worth memoizing!
+// this needs even more time to calculate because of the overhead of memoizing, small things are not worth memoizing
+const doubler = useCallback(
+  () => x * 2,
+  [x], 
+);
+```
+
+### [useMemo](https://reactjs.org/docs/hooks-reference.html#usememo)
+
+`useMemo` is the same as `useCallback` but instead of returning a memoized callback, it returns a memoized value. Similarly to `useCallback`, you should only use this for things that are expensive
```

Comment:
I'd add `useMemo` is the equivalent of 
```
useCallback(() => {
  console.log('foo')
}, [])();
```

Reply:
Good idea, I wrote up a bit of a more in depth explanation about the similarities between the hooks that uses that example